### DISCUSSION TITLE
How "exactly" hybrid search works?

### DISCUSSION THREAD

**Discussion by SeunghyunYoo** [milvus-io/milvus#26841] [Q&A and General discussion]
I want to know how exactly hybrid search works.

According to previous `Hybrid Search Behaviour` [discussion](https://github.com/milvus-io/milvus/discussions/12927?sort=new), hybrid search works through followed step.
> first, filtering vectors by the expression condition.
 second, do ANN search in the vectors which is filtered out, return topk items

However, it seems like it works differently.

I tested hybrid search with `IVF_FLAT` index and `95752` items.

The item contains `item_id, item_vector, service` fields. The item meta info (`item count per service`) is followed.

```
blog: 62092
bboom: 2836
influencer_topic: 7550
post: 11423
tv: 6260
premium_content: 5590
audioclip: 1
```

---

Here's my testing result.

I create `IVF_FLAT` index on `item_vector` with `1236 nlist`
```python
fields = [
    FieldSchema(name="item_id", dtype=DataType.VARCHAR, is_primary=True, auto_id=False, max_length=100),
    FieldSchema(name="item_vector", dtype=DataType.FLOAT_VECTOR, dim=dim),
    FieldSchema(name="service", dtype=DataType.VARCHAR, max_length=100),
]

collection.create_index(
    "item_vector",
    {
        "index_type": "IVF_FLAT",
        "metric_type": "IP",
        "params": {"nlist": 1236}
    }
)
collection.create_index(
    field_name="service",
    index_name="service_index",
)
```

And I checked collection contains `audioclip` data.
```python
result = collection.query(
    expr='service == "audioclip"',
    offset=0,
    limit=10,
    output_fields=["item_id", "service"],
)

print(result)
```
> [{'item_id': 'ACLIP_C:261_283', 'service': 'audioclip'}]

---

The result was different depends on `nprobe` parameter.

#### case 1: nprobe = 256
```python
search_params = {
    "metric_type": "IP",
    "params": {"nprobe": 256},
}

result = collection.search(
    user_vectors[0].reshape(1, -1), "item_vector", search_params, 200,
    consistency_level=args.consistency_level,
    expr=f'service == "audioclip"'
)
result = [(item_id, score) for item_id, score in zip(result[0].ids, result[0].distances)]

print(result)
```
> nothing

#### case 2: nprobe = 1000
```python
search_params = {
    "metric_type": "IP",
    "params": {"nprobe": 1000},
}

result = collection.search(
    user_vectors[0].reshape(1, -1), "item_vector", search_params, 200,
    consistency_level=args.consistency_level,
    expr=f'service == "audioclip"'
)
result = [(item_id, score) for item_id, score in zip(result[0].ids, result[0].distances)]

print(result)
```
> [('ACLIP_C:261_283', -0.577196478843689)]

---

Based on my testing result, I assume hybrid search works through followed steps.
- extract centroids. the amount of centroids is same as `nprobe size`
- then, aggregate filtering with `service == "audioclip"` expression on the extracted clusters
- finally, ANN similarity search

If there was no `audioclip` data in 256 nearest clusters data, the hybrid search result (w/ `256` nprobe) could be nothing.
The hybrid search result (w/ `1000` nprobe) has 1 data, because there is `audioclip` data somewhere in `1000 nearest clusters`

---

So, I want to know `how exactly hybrid search works`.

I also want to know hybrid search mechanism depend on `index type`.
In my test scenario, the search mechanism looks depending on `index type`.

**Comment by yhmo** [comment]
The actual steps:
- filtering with service == "audioclip", build a bitset for all the items. For example, a segment has 5 items, only the second item matches the condition, the bitset for this segment is [0, 1, 0, 0, 0]
- the index has nlist centroids, now we search for nprobe centroids.
- calculate distance in the nproble centroids, find out the topk nearest items

Assume a segment has 10000 items, nlist=100, there are 100 buckets in the index. But there is only 1 item that matches the condition "service == "audioclip"", this item is in the No.50 bucket. Now I search with nprobe=5, unfortunately, the 5 buckets are No.3, No.4, No.5, No.6, No.7, then the result is empty. Then I search again with nprobe=50, the No.50 bucket is included, the result return the one item for me.

**Reply by SeunghyunYoo** [reply]
@yhmo,
Thanks for your fast answer :). I have another question.

I read milvus paper right before.
- https://www.cs.purdue.edu/homes/csjgwang/pubs/SIGMOD21_Milvus.pdf

milvus choose search strategy by cost optimizer

your explanation is same as `Strategy B: attribute-first-vector-search.`
> The difference with the strategy A is that after it obtains the relevant entities according to attribute constraint C𝐴, it produces a bitmap of the resultant entity IDs.
Then it conducts the normal vector query processing based on C𝑉 and checks the bitmap whenever a vector is encountered.
Only vectors that pass bitmap testing are included in the final top-𝑘 results.
This strategy is suitable in many cases when C𝐴 or C𝑉 is moderately selective.

I also think milvus probably choose `Strategy C: vector-first-attribute-full-scan` in the testing result, because the selectivity of service fields is low (`6 / 95752`).
> In contrast to the strategy A, this approach only uses the vector constraint C𝑉 to obtain the relevant entities via vector indexing like IVF_FLAT.
Then the resultant entities are fully scanned to verify if they satisfy the attribute constraint C𝐴. To make sure there are 𝑘 final results, it searches for 𝜃 · 𝑘 (𝜃 > 1) results during the vector query processing.
This strategy is suitable when the vector constraint C𝑉 is highly selective that the number of candidates is relatively small.

To be chosen `strategy B`, milvus have to conclude C𝐴 has quite good selectivity.
That is, milvus have to save another stats meta like how many data exists per service fields.
like this.
```
blog: 62092
bboom: 2836
influencer_topic: 7550
post: 11423
tv: 6260
premium_content: 5590
audioclip: 1
```

Dose milvus save the another stats meta?

**Reply by yhmo** [reply]
The paper describes the possible solutions for vector hybrid search.
A:  Attribute filtering  ---> Vector full-scan(FLAT)
B: Attribute filtering ---> Index search (IVF, HNSW, etc)
C: Index search (IVF, HNSW, etc) ---> Attribute filtering
D: Cost-based planning
E: Partition-based filtering
.....

In fact, "vector full-scan" is brute-search without index. In Milvus, if you call create_index(index_type="FLAT"), that is a full scan. So, both the A and B are supported by Milvus in current versions. 

We don't think C is a good solution because it is hard to estimate how many vectors should be searched to ensure the topk results. On the other hand, by A and B, we can possibly ensure topk results. For example, in HNSW, we can travel more graph nodes if the results are not reached topk.

B doesn't require extra metadata, we just scan the attribute fields by the filtering condition to construct the bitmap. But it is possible to add some stats to accelerate filtering speed(we do have a plan to do this).

**Reply by SeunghyunYoo** [reply]
thanks :)

I thought your team publish that papers, but it isn't.

> milvus have to save another stats meta like how many data exists per service fields.

This one is based on my thought that milvus choose search strategy at runtime or update strategy on index periodically.

---

> it is hard to estimate how many vectors should be searched to ensure the topk results.

I agree with this opinion. 

I totally understand how hybrid search works now. thanks again :)

**Reply by yhmo** [reply]
> I thought your team publish that papers, but it isn't.

We did publish the paper. My name is also on the name list, but I am just not the major researcher.

**Comment by codergma** [comment]
I think this is a great discussion, and I've learned a lot.
